src/coreclr/jit/optimizer.cpp

```diff
@@ -6205,42 +6236,96 @@ void Compiler::optHoistThisLoop(unsigned lnum, LoopHoistContext* hoistCtxt)
     // or side-effect dependent things.
     //
     // We really should consider hoisting from conditionally executed blocks, if they are frequently executed
-    // and it is safe to evaluate the tree early.
-    //
-    // In particular if we have a loop nest, when scanning the outer loop we should consider hoisting from blocks
-    // in enclosed loops. However, this is likely to scale poorly, and we really should instead start
-    // hoisting inner to outer.
+    // and it is safe to evaluate the tree early
     //
     ArrayStack<BasicBlock*> defExec(getAllocatorLoopHoist());
+
+    bool pushAllPreheaders = false;
+
     if (pLoopDsc->lpExitCnt == 1)
     {
         assert(pLoopDsc->lpExit != nullptr);
-        JITDUMP("  Only considering hoisting in blocks that dominate exit block " FMT_BB "\n", pLoopDsc->lpExit->bbNum);
-        BasicBlock* cur = pLoopDsc->lpExit;
+        JITDUMP("  Considering hoisting in blocks that either dominate exit block " FMT_BB
+                " or preheaders of nested loops, if any:\n",
+                pLoopDsc->lpExit->bbNum);
+
         // Push dominators, until we reach "entry" or exit the loop.
+        // Also push the preheaders that were added for the nested loops,
+        // if any, along the way such that the final block list is arranged
+        // in execution order
```

Comment:
For the dominator-walk added blocks calling this "execution order" makes sense. 

But for the child preheaders, less so.

I think you are trying to say that the order of the blocks you end up with is consistent with a dominator tree preorder.

I wonder if we should just create preheaders up front before building dominators, so that we don't have to do all this extra work to figure out how they fit in.

Reply:
> I wonder if we should just create preheaders up front before building dominators

Bruce tried to do it but ran into some problems. I will follow up with him to revamp that work, but if it gets too complicated, I could do it in a follow-up PR.

Reply:
I tried to force create loop preHeaders, but there are lot of design problems that needs to be addressed. Currently, they are created only during loop hoisting and other loop optimizations like cloning, unrolling, etc. are not aware of them. Once we start creating preheaders, their state needs to be taken into account or updated when we do these loop optimizations. I was debugging a case where we create a loop preheader and then clone the loop. In such cases, we just update the loop head in `optUpdateLoopHead()` to newer head, but we do not take into account if it was marked as prehead. In such circumstances, we should ideally create a new loop preheader and unmark the existing preheader. 

![image](https://user-images.githubusercontent.com/12488060/172878106-8416f9e2-4647-4bf2-b7c5-e206cf3f1e8d.png)

There might be similar complexities involved while unrolling the loop. Not that none of this is not doable, but I think it is needs some design thinking and might not be worth doing as part of this PR. 

I also realized that if I do preheader, it changes the bbNum of blocks. And even if nothing else changes, it definitely impacts the LSRA block sequencing which triggers lot of diffs.

On the side note, in this PR, I am conservative where I check if a sibling of a node can throw exception and if yes, do not bother hoisting it. Sometimes this might negative impact us. For instance, in one of the example, we do not hoist the array access because of presence of bound check, but then later, we eliminate that very bound check through assertion prop, but we didn't have that information earlier. But from correctness perspective, we should continue to be conservative in these things.

Reply:
Ah, didn't realize you suggested to rephrase. Done.